target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c

```diff
@@ -0,0 +1,506 @@
+ */
+
+static const struct clk_ops rtl_clk_ops = {
+	.round_rate = rtl_round_rate,
+	.recalc_rate = rtl_recalc_rate,
+	.set_rate = rtl_set_rate,
+};
+
+static int rtl_ccu_create(struct device_node *np)
+{
+	if (soc_info.family != RTL8380_FAMILY_ID && soc_info.family != RTL8390_FAMILY_ID)
+		return -ENXIO;
+
+	rtl_ccu = kzalloc(sizeof(*rtl_ccu), GFP_KERNEL);
+	if (IS_ERR(rtl_ccu))
+		return -ENOMEM;
+
+	rtl_ccu->np = np;
+	spin_lock_init(&rtl_ccu->lock);
+
+	switch (soc_info.family) {
+	case RTL8380_FAMILY_ID:
+		rtl_ccu->soc = SOC_RTL838X;
+		break;
+	case RTL8390_FAMILY_ID:
+		rtl_ccu->soc = SOC_RTL839X;
+		break;
+	}
+
+	return 0;
+}
+
+int rtl_clk_hw_register(int idx)
+{
+	struct clk_parent_data parent_data = { };
+	struct clk_init_data hw_init = { };
+	struct rtl_clk *clk;
+	int ret;
+
+	clk = &rtl_ccu->clks[idx];
+
+	clk->hw.init = &hw_init;
+	clk->idx = idx;
+
+	parent_data.fw_name = rtl_clk_info[idx].parent_name;
+
+	hw_init.name = rtl_clk_info[idx].name;
+	hw_init.ops = &rtl_clk_ops;
+	hw_init.parent_data = &parent_data;
+	hw_init.num_parents = 1;
+
+	ret = of_clk_hw_register(rtl_ccu->np, &clk->hw);
+	if (ret)
+		return ret;
+
+	clk_hw_register_clkdev(&clk->hw, rtl_clk_info[idx].name, NULL);
+
+	return 0;
+}
+
+static struct clk_hw *rtl_clk_hw_get(struct of_phandle_args *clkspec, void *priv)
+{
+	unsigned int idx = clkspec->args[0];
+
+	if (idx >= CLK_COUNT) {
+		pr_err("%s: Invalid index %u\n", __func__, idx);
+		return ERR_PTR(-EINVAL);
+	}
+
+	return &rtl_ccu->clks[idx].hw;
+}
+
+static int rtl_ccu_register_clocks(void)
+{
+	int idx, ret;
+
+	for (idx = 0; idx < CLK_COUNT; ++idx) {
+		ret = rtl_clk_hw_register(idx);
+		if (ret) {
+			pr_err("%s: Couldn't register %s clock\n",
+				__func__, rtl_clk_info[idx].display_name);
+			goto err_hw_unregister;
+		}
+	}
+
+	ret = of_clk_add_hw_provider(rtl_ccu->np, rtl_clk_hw_get, rtl_ccu);
+	if (ret) {
+		pr_err("%s: Couldn't register clock provider of %s\n",
+			__func__, of_node_full_name(rtl_ccu->np));
+		goto err_hw_unregister;
+	}
+
+	return 0;
+
+err_hw_unregister:
+	for (--idx; idx >= 0; --idx)
+		clk_hw_unregister(&rtl_ccu->clks[idx].hw);
+
+	return ret;
+}
+
+void rtl_ccu_log(void)
+{
+	int idx, rate;
+	struct clk *clk;
+	char msg[255], msgl[255] = "RTL83XX clocks initialized";
+
+	for (idx = 0; idx < CLK_COUNT; idx++) {
+		clk = clk_get(NULL, rtl_clk_info[idx].name);
+		rate = clk_get_rate(clk);
+		clk_put(clk);
+
+		sprintf(msg, ", %s %d MHz", rtl_clk_info[idx].display_name, rate / 1000000);
+		strcat(msgl, msg);
+	}
+	pr_info("%s\n", msgl);
+}
+
+static void __init rtl_clk_init(struct device_node *np)
+{
+	if (rtl_ccu_create(np))
+		return;
+
+	if (rtl_ccu_register_clocks())
+		kfree(rtl_ccu);
+	else
+		rtl_ccu_log();
+}
+
+/* 
+ * This module provides early startup clocks that are needed for further
+ * devices (e.g. UART). Therefore we cannot use the platform driver framework.
+ */
+
+CLK_OF_DECLARE(rtl83xx, "realtek,rtl83xx-clock", rtl_clk_init);
```

Comment:
```c
CLK_OF_DECLARE(rtl838x, "realtek,rtl838x-clock", rtl_clk_init);
CLK_OF_DECLARE(rtl839x, "realtek,rtl839x-clock", rtl_clk_init);
```
The rtl838x/rtl839x names are likely to be refused upstream though. You could use rtl8380/rtl8390 to classify these peripherals, which is what I used for the GPIO and watchdog drivers. Alternatively, you could also use Realtek's codenames "maple" and "cypress", although those make it harder to identify which drivers certain hardware would require.

Reply:
fixed in upcoming version